File: csharp/src/DarkStrata.CredentialCheck/Crypto/CryptoUtils.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;
#if NETSTANDARD2_0
using DarkStrata.CredentialCheck.Compatibility;
#endif

namespace DarkStrata.CredentialCheck;

/// <summary>
/// Cryptographic utility functions for credential checking.
/// </summary>
#if NETSTANDARD2_0
public static class CryptoUtils
#else
public static partial class CryptoUtils
#endif
{
    private const int PrefixLength = 5;

#if NETSTANDARD2_0
    private static readonly Regex HexPatternRegex = new Regex("^[A-Fa-f0-9]+$", RegexOptions.Compiled);
#endif

    /// <summary>
    /// Compute SHA-256 hash of a credential pair.
    /// The credential is formatted as "email:password" before hashing.
    /// The email is normalized (lowercased and trimmed) before hashing.
    /// </summary>
    /// <param name="email">The email address or username.</param>
    /// <param name="password">The password.</param>
    /// <returns>The SHA-256 hash as an uppercase hexadecimal string.</returns>
    public static string HashCredential(string email, string password)
    {
        var normalizedEmail = email.Trim().ToLowerInvariant();
        var credential = $"{normalizedEmail}:{password}";
        return Sha256(credential);
    }

    /// <summary>
    /// Compute SHA-256 hash of a string.
    /// </summary>
    /// <param name="input">The string to hash.</param>
    /// <returns>The SHA-256 hash as an uppercase hexadecimal string.</returns>
    public static string Sha256(string input)
    {
        var bytes = Encoding.UTF8.GetBytes(input);
#if NETSTANDARD2_0
        var hash = PolyfillHelpers.Sha256Hash(bytes);
        return PolyfillHelpers.ToHexString(hash);
#else
        var hash = SHA256.HashData(bytes);
        return Convert.ToHexString(hash);
#endif
    }

    /// <summary>
    /// Compute HMAC-SHA256 of a message with a key.
    /// </summary>
    /// <param name="message">The message to authenticate.</param>
    /// <param name="keyHex">The HMAC key as a hexadecimal string.</param>
    /// <returns>The HMAC-SHA256 as an uppercase hexadecimal string.</returns>
    public static string HmacSha256(string message, string keyHex)
    {
#if NETSTANDARD2_0
        var keyBytes = PolyfillHelpers.FromHexString(keyHex);
        var messageBytes = Encoding.UTF8.GetBytes(message);
        var hmac = PolyfillHelpers.HmacSha256Hash(keyBytes, messageBytes);
        return PolyfillHelpers.ToHexString(hmac);
#else
        var keyBytes = Convert.FromHexString(keyHex);
        var messageBytes = Encoding.UTF8.GetBytes(message);
        var hmac = HMACSHA256.HashData(keyBytes, messageBytes);
        return Convert.ToHexString(hmac);
#endif
    }

    /// <summary>
    /// Extract the k-anonymity prefix from a hash.
    /// </summary>
    /// <param name="hash">The full SHA-256 hash (64 hex characters).</param>
    /// <returns>The first 5 characters (prefix) in uppercase.</returns>
    public static string ExtractPrefix(string hash)
    {
        return hash.Substring(0, PrefixLength).ToUpperInvariant();
    }

    /// <summary>
    /// Check if a hash is in a set of HMAC'd hashes.
    /// Uses timing-safe comparison to prevent timing attacks.
    /// </summary>
    /// <param name="hash">The full hash to check.</param>
    /// <param name="hmacKey">The HMAC key from the API response.</param>
    /// <param name="hmacHashes">Collection of HMAC'd hashes from the API.</param>
    /// <returns>True if the hash is found in the set.</returns>
    public static bool IsHashInSet(string hash, string hmacKey, IEnumerable<string> hmacHashes)
    {
        // Compute HMAC of the full hash
        var targetHmac = HmacSha256(hash, hmacKey);
#if NETSTANDARD2_0
        var targetBytes = PolyfillHelpers.FromHexString(targetHmac);
#else
        var targetBytes = Convert.FromHexString(targetHmac);
#endif

        foreach (var hmacHash in hmacHashes)
        {
            try
            {
#if NETSTANDARD2_0
                var candidateBytes = PolyfillHelpers.FromHexString(hmacHash);
                if (targetBytes.Length == candidateBytes.Length &&
                    PolyfillHelpers.FixedTimeEquals(targetBytes, candidateBytes))
                {
                    return true;
                }
#else
                var candidateBytes = Convert.FromHexString(hmacHash);
                if (targetBytes.Length == candidateBytes.Length &&
                    CryptographicOperations.FixedTimeEquals(targetBytes, candidateBytes))
                {
                    return true;
                }
#endif
            }
            catch (FormatException)
            {
                // Invalid hex string, skip
                continue;
            }
        }

        return false;
    }

    /// <summary>
    /// Validate that a string is a valid hexadecimal hash.
    /// </summary>
    /// <param name="hash">The string to validate.</param>
    /// <param name="expectedLength">Expected length (default: 64 for SHA-256).</param>
    /// <returns>True if the string is valid hex of the expected length.</returns>
    public static bool IsValidHash(string hash, int expectedLength = 64)
    {
        if (hash.Length != expectedLength)
        {
            return false;
        }
#if NETSTANDARD2_0
        return HexPatternRegex.IsMatch(hash);
#else
        return HexPattern().IsMatch(hash);
#endif
    }

    /// <summary>
    /// Validate that a string is a valid k-anonymity prefix.
    /// </summary>
    /// <param name="prefix">The prefix to validate.</param>
    /// <returns>True if the prefix is valid (5 hex characters).</returns>
    public static bool IsValidPrefix(string prefix)
    {
#if NETSTANDARD2_0
        return prefix.Length == PrefixLength && HexPatternRegex.IsMatch(prefix);
#else
        return prefix.Length == PrefixLength && HexPattern().IsMatch(prefix);
#endif
    }

    /// <summary>
    /// Group credentials by their hash prefix for efficient batch processing.
    /// </summary>
    /// <typeparam name="T">The credential type (must have a Hash property).</typeparam>
    /// <param name="credentials">Array of credential objects with Hash property.</param>
    /// <param name="hashSelector">Function to extract the hash from a credential.</param>
    /// <returns>Dictionary of prefix to list of credentials.</returns>
    public static Dictionary<string, List<T>> GroupByPrefix<T>(
        IEnumerable<T> credentials,
        Func<T, string> hashSelector)
    {
        var groups = new Dictionary<string, List<T>>();

        foreach (var credential in credentials)
        {
            var prefix = ExtractPrefix(hashSelector(credential));
            if (!groups.TryGetValue(prefix, out var list))
            {
                list = new List<T>();
                groups[prefix] = list;
            }
            list.Add(credential);
        }

        return groups;
    }

#if !NETSTANDARD2_0
    [System.Text.RegularExpressions.GeneratedRegex("^[A-Fa-f0-9]+$")]
    private static partial Regex HexPattern();
#endif
}

File: sdks/java/src/main/java/io/darkstrata/credentialcheck/CryptoUtils.java
package io.darkstrata.credentialcheck;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.InvalidKeyException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Cryptographic utilities for credential checking.
 */
public final class CryptoUtils {

    private static final Pattern HEX_PATTERN = Pattern.compile("^[0-9A-Fa-f]+$");
    private static final char[] HEX_ARRAY = "0123456789ABCDEF".toCharArray();

    private CryptoUtils() {
        // Utility class
    }

    /**
     * Hash a credential (email:password) using SHA-256.
     * The email is normalized (lowercased and trimmed) before hashing.
     *
     * @param email    the email address
     * @param password the password
     * @return uppercase hex-encoded SHA-256 hash
     */
    public static String hashCredential(String email, String password) {
        String normalizedEmail = email.trim().toLowerCase(Locale.ROOT);
        return sha256(normalizedEmail + ":" + password);
    }

    /**
     * Compute SHA-256 hash of the input string.
     *
     * @param input the input string
     * @return uppercase hex-encoded SHA-256 hash
     */
    public static String sha256(String input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
            return bytesToHex(hash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 algorithm not available", e);
        }
    }

    /**
     * Compute HMAC-SHA256 of a message with a hex-encoded key.
     *
     * @param message the message to authenticate
     * @param keyHex  the key in hex format
     * @return uppercase hex-encoded HMAC
     */
    public static String hmacSha256(String message, String keyHex) {
        try {
            byte[] keyBytes = hexToBytes(keyHex);
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(keyBytes, "HmacSHA256");
            mac.init(secretKey);
            byte[] hmacBytes = mac.doFinal(message.getBytes(StandardCharsets.UTF_8));
            return bytesToHex(hmacBytes);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("HmacSHA256 algorithm not available", e);
        } catch (InvalidKeyException e) {
            throw new RuntimeException("Invalid HMAC key", e);
        }
    }

    /**
     * Extract the prefix (first 5 characters) from a hash.
     *
     * @param hash the full hash
     * @return the 5-character prefix in uppercase
     */
    public static String extractPrefix(String hash) {
        if (hash == null || hash.length() < Constants.PREFIX_LENGTH) {
            throw new IllegalArgumentException("Hash must be at least " + Constants.PREFIX_LENGTH + " characters");
        }
        return hash.substring(0, Constants.PREFIX_LENGTH).toUpperCase(Locale.ROOT);
    }

    /**
     * Validate that a string is a valid hex hash of expected length.
     *
     * @param hash           the hash to validate
     * @param expectedLength the expected length (default 64 for SHA-256)
     * @return true if valid
     */
    public static boolean isValidHash(String hash, int expectedLength) {
        if (hash == null || hash.length() != expectedLength) {
            return false;
        }
        return HEX_PATTERN.matcher(hash).matches();
    }

    /**
     * Validate that a string is a valid SHA-256 hash (64 hex characters).
     */
    public static boolean isValidHash(String hash) {
        return isValidHash(hash, 64);
    }

    /**
     * Validate that a string is a valid hash prefix (5 hex characters).
     */
    public static boolean isValidPrefix(String prefix) {
        if (prefix == null || prefix.length() != Constants.PREFIX_LENGTH) {
            return false;
        }
        return HEX_PATTERN.matcher(prefix).matches();
    }

    /**
     * Check if a hash is in a set of HMAC'd hashes using timing-safe comparison.
     *
     * @param hash       the full credential hash
     * @param hmacKey    the HMAC key in hex
     * @param hmacHashes the list of HMAC'd hashes from the API
     * @return true if the hash is found in the set
     */
    public static boolean isHashInSet(String hash, String hmacKey, List<String> hmacHashes) {
        String hmacOfHash = hmacSha256(hash.toUpperCase(Locale.ROOT), hmacKey);

        // Timing-safe comparison - always compare against all hashes
        boolean found = false;
        for (String hmacHash : hmacHashes) {
            if (timingSafeEquals(hmacOfHash, hmacHash.toUpperCase(Locale.ROOT))) {
                found = true;
                // Don't break early - continue to prevent timing attacks
            }
        }
        return found;
    }

    /**
     * Group credentials by their hash prefix.
     *
     * @param credentials list of credentials with computed hashes
     * @return map of prefix to list of credentials
     */
    public static <T> Map<String, List<T>> groupByPrefix(List<T> credentials, java.util.function.Function<T, String> hashExtractor) {
        Map<String, List<T>> groups = new HashMap<>();
        for (T credential : credentials) {
            String hash = hashExtractor.apply(credential);
            String prefix = extractPrefix(hash);
            groups.computeIfAbsent(prefix, k -> new ArrayList<>()).add(credential);
        }
        return groups;
    }

    /**
     * Convert bytes to uppercase hex string.
     */
    private static String bytesToHex(byte[] bytes) {
        char[] hexChars = new char[bytes.length * 2];
        for (int i = 0; i < bytes.length; i++) {
            int v = bytes[i] & 0xFF;
            hexChars[i * 2] = HEX_ARRAY[v >>> 4];
            hexChars[i * 2 + 1] = HEX_ARRAY[v & 0x0F];
        }
        return new String(hexChars);
    }

    /**
     * Convert hex string to bytes.
     */
    private static byte[] hexToBytes(String hex) {
        int len = hex.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(hex.charAt(i), 16) << 4)
                    + Character.digit(hex.charAt(i + 1), 16));
        }
        return data;
    }

    /**
     * Timing-safe string comparison.
     */
    private static boolean timingSafeEquals(String a, String b) {
        if (a == null || b == null) {
            return false;
        }
        if (a.length() != b.length()) {
            return false;
        }

        int result = 0;
        for (int i = 0; i < a.length(); i++) {
            result |= a.charAt(i) ^ b.charAt(i);
        }
        return result == 0;
    }
}
